### PR TITLE
Improve checking table privileges in sql-standard access control

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePrincipal.java [hunk 1]
```diff
@@ -27,6 +29,16 @@
 
 public class HivePrincipal
 {
+    public static HivePrincipal from(ConnectorIdentity identity)
+    {
+        if (!identity.getRole().isPresent() || identity.getRole().get().getType() == SelectedRole.Type.ALL) {
+            return new HivePrincipal(PrincipalType.USER, identity.getUser());
+        }
+        else {
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePrincipal.java]: redundant `else`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java [hunk 2]
```diff
@@ -316,13 +316,7 @@ public static boolean isRoleEnabled(ConnectorIdentity identity, Function<HivePri
...
         }
 
-        HivePrincipal principal;
-        if (!identity.getRole().isPresent() || identity.getRole().get().getType() == SelectedRole.Type.ALL) {
-            principal = new HivePrincipal(USER, identity.getUser());
-        }
-        else {
-            principal = new HivePrincipal(ROLE, identity.getRole().get().getRole().get());
-        }
+        HivePrincipal principal = HivePrincipal.from(identity);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java]: is this new method applicable here too?
https://github.com/prestosql/presto/blob/c672138fde86472ab470aaa7943f61e010387007/presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java#L347-L353

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 3]
```diff
@@ -825,17 +825,18 @@ public synchronized void revokeRoles(Set<String> roles, Set<HivePrincipal> grant
         }
         switch (tableAction.getType()) {
             case ADD:
-            case ALTER:
-                if (principal.getType() == PrincipalType.ROLE) {
+            case ALTER: {
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: `{` looks redundant or even unmatched

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/PermissionMetadata.java [hunk 4]
```diff
@@ -54,8 +56,14 @@ public boolean isGrantOption()
...
+    @JsonProperty
+    public HivePrincipal getGrantee()
+    {
+        return grantee;
+    }
+
     public HivePrivilegeInfo toHivePrivilegeInfo()
     {
-        return new HivePrivilegeInfo(permission, grantOption, new HivePrincipal(USER, "admin"), new HivePrincipal(USER, "admin"));
+        return new HivePrivilegeInfo(permission, grantOption, new HivePrincipal(USER, "admin"), grantee);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/PermissionMetadata.java]: im not convinced this belongs to `Make principal in listTablePrivileges Optional` commit

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 5]
```diff
@@ -432,7 +433,9 @@ private boolean checkTablePermission(
         }
 
         SemiTransactionalHiveMetastore metastore = metastoreProvider.apply(((HiveTransactionHandle) context.getTransactionHandle()));
-        return listEnabledTablePrivileges(metastore, tableName.getSchemaName(), tableName.getTableName(), context.getIdentity())
+        return metastore.listTablePrivileges(new HiveIdentity(context.getIdentity()), tableName.getSchemaName(), tableName.getTableName(), Optional.empty())
+                .stream()
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java]: put .stream on prev line

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 6]
```diff
@@ -432,7 +433,9 @@ private boolean checkTablePermission(
...
+                .filter(privilegeInfo -> listEnabledPrincipals(metastore, context.getIdentity()).anyMatch(privilegeInfo.getGrantee()::equals))
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java]: line break before .anyMatch

(can be also line break after ->)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 7]
```diff
@@ -481,7 +484,7 @@ private boolean hasAnyTablePermission(ConnectorSecurityContext context, SchemaTa
         }
 
         SemiTransactionalHiveMetastore metastore = metastoreProvider.apply(((HiveTransactionHandle) context.getTransactionHandle()));
-        return listEnabledTablePrivileges(metastore, tableName.getSchemaName(), tableName.getTableName(), context.getIdentity())
-                .anyMatch(privilegeInfo -> true);
+        return metastore.listTablePrivileges(new HiveIdentity(context.getIdentity()), tableName.getSchemaName(), tableName.getTableName(), Optional.empty()).stream()
+                .anyMatch(privilegeInfo -> listEnabledPrincipals(metastore, context.getIdentity()).anyMatch(privilegeInfo.getGrantee()::equals));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java]: line break before .anyMatch

(can be also line break after ->)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 8]
```diff
@@ -484,7 +490,12 @@ private boolean hasAnyTablePermission(ConnectorSecurityContext context, SchemaTa
         }
 
         SemiTransactionalHiveMetastore metastore = metastoreProvider.apply(((HiveTransactionHandle) context.getTransactionHandle()));
-        return metastore.listTablePrivileges(new HiveIdentity(context.getIdentity()), tableName.getSchemaName(), tableName.getTableName(), Optional.empty()).stream()
-                .anyMatch(privilegeInfo -> listEnabledPrincipals(metastore, context.getIdentity()).anyMatch(privilegeInfo.getGrantee()::equals));
+        Set<HivePrincipal> allowedPrincipals = metastore.listTablePrivileges(new HiveIdentity(context.getIdentity()), tableName.getSchemaName(), tableName.getTableName(), Optional.empty())
+                .stream()
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java]: prev line

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 9]
```diff
@@ -433,11 +435,15 @@ private boolean checkTablePermission(
         }
 
         SemiTransactionalHiveMetastore metastore = metastoreProvider.apply(((HiveTransactionHandle) context.getTransactionHandle()));
-        return metastore.listTablePrivileges(new HiveIdentity(context.getIdentity()), tableName.getSchemaName(), tableName.getTableName(), Optional.empty())
-                .stream()
-                .filter(privilegeInfo -> listEnabledPrincipals(metastore, context.getIdentity()).anyMatch(privilegeInfo.getGrantee()::equals))
+
+        Set<HivePrincipal> allowedPrincipals = metastore.listTablePrivileges(new HiveIdentity(context.getIdentity()), tableName.getSchemaName(), tableName.getTableName(), Optional.empty()).stream()
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java]: Include rationale in `Do not enumerate all enabled roles when checking access in sql-standard` commit

**Comment by kokosing** [comment]
Applied comments.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java [hunk 10]
```diff
@@ -107,7 +107,10 @@
 
     void revokeTablePrivileges(String databaseName, String tableName, String tableOwner, HivePrincipal grantee, Set<HivePrivilegeInfo> privileges);
 
-    Set<HivePrivilegeInfo> listTablePrivileges(String databaseName, String tableName, String tableOwner, HivePrincipal principal);
+    /**
+     * @param principal when empty, all table privileges are returned
+     */
+    Set<HivePrivilegeInfo> listTablePrivileges(String databaseName, String tableName, String tableOwner, Optional<HivePrincipal> principal);
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java]: I don't like. One can say that we could put `for` before any parameter name.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 11]
```diff
@@ -979,21 +981,36 @@ private static boolean partitionMatches(String partitionName, List<String> parts
...
-            result.add(new HivePrivilegeInfo(OWNERSHIP, true, principal, principal));
+        if (principal.get().getType() == USER && table.getOwner().equals(principal.get().getName())) {
+            result.add(new HivePrivilegeInfo(OWNERSHIP, true, principal.get(), principal.get()));
         }
-        result.addAll(readPermissionsFile(getPermissionsPath(permissionsDirectory, principal)));
+        result.addAll(readPermissionsFile(getPermissionsPath(permissionsDirectory, principal.get())));
         return result.build();
     }
 
+    private List<Path> listFiles(Path directory)
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java]: This some kind of a rebase mistake. Thanks.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java]: This one is really ambiguous.

[the remark this answers is not in the document]
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java]: At least it used to be ambiguous for me.

**Pull request merged**